node/pkg/query/response.go

```diff
@@ -0,0 +1,507 @@
+	QueryFatalError QueryStatus = -1
+)
+
+// This is the query response returned from the watcher to the query handler.
+type PerChainQueryResponseInternal struct {
+	RequestID  string
+	RequestIdx int
+	ChainId    vaa.ChainID
+	Status     QueryStatus
+	Response   ChainSpecificResponse
+}
+
+// CreatePerChainQueryResponseInternal creates a PerChainQueryResponseInternal and returns a pointer to it.
+func CreatePerChainQueryResponseInternal(reqId string, reqIdx int, chainId vaa.ChainID, status QueryStatus, response ChainSpecificResponse) *PerChainQueryResponseInternal {
+	return &PerChainQueryResponseInternal{
+		RequestID:  reqId,
+		RequestIdx: reqIdx,
+		ChainId:    chainId,
+		Status:     status,
+		Response:   response,
+	}
+}
+
+var queryResponsePrefix = []byte("query_response_0000000000000000000|")
+
+// QueryResponsePublication is the response to a QueryRequest.
+type QueryResponsePublication struct {
+	Request           *gossipv1.SignedQueryRequest
+	PerChainResponses []*PerChainQueryResponse
+}
+
+// PerChainQueryResponse represents a query response for a single chain.
+type PerChainQueryResponse struct {
+	// ChainId indicates which chain this query was destine for.
+	ChainId vaa.ChainID
+
+	// Response is the chain specific query data.
+	Response ChainSpecificResponse
+}
+
+// ChainSpecificResponse is the interface that must be implemented by a chain specific response.
+type ChainSpecificResponse interface {
+	Type() ChainSpecificQueryType
+	Marshal() ([]byte, error)
+	Unmarshal(data []byte) error
+	UnmarshalFromReader(reader *bytes.Reader) error
+	Validate() error
+}
+
+// EthCallQueryResponse implements ChainSpecificResponse for an EVM eth_call query response.
+type EthCallQueryResponse struct {
+	BlockNumber uint64
+	Hash        common.Hash
+	Time        time.Time
+
+	// Results is the array of responses matching CallData in EthCallQueryRequest
+	Results [][]byte
+}
+
+//
+// Implementation of QueryResponsePublication.
+//
+
+// Marshal serializes the binary representation of a query response.
+// This method calls Validate() and relies on it to range checks lengths, etc.
+func (msg *QueryResponsePublication) Marshal() ([]byte, error) {
+	if err := msg.Validate(); err != nil {
+		return nil, err
+	}
+
+	buf := new(bytes.Buffer)
+
+	vaa.MustWrite(buf, binary.BigEndian, uint8(1)) // version
+
+	// Source
+	// TODO: support writing off-chain and on-chain requests
+	// Here, unset represents an off-chain request
+	vaa.MustWrite(buf, binary.BigEndian, vaa.ChainIDUnset)
+
+	buf.Write(msg.Request.Signature[:])
+
+	// Write the length of the request to facilitate on-chain parsing.
+	if len(msg.Request.QueryRequest) > math.MaxUint32 {
+		return nil, fmt.Errorf("request too long")
+	}
+	vaa.MustWrite(buf, binary.BigEndian, uint32(len(msg.Request.QueryRequest)))
+
+	buf.Write(msg.Request.QueryRequest)
+
+	// Per chain responses
+	vaa.MustWrite(buf, binary.BigEndian, uint8(len(msg.PerChainResponses)))
```

Comment:
should this uint8 be explicitly checked? (can be done in a follow up PR)

Reply:
This is checked in `Validate` which is called at the start of this function.